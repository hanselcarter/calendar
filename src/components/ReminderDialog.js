import React from "react";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogContentText from "@material-ui/core/DialogContentText";
import DialogTitle from "@material-ui/core/DialogTitle";
import PropTypes from "prop-types";
import DatePicker from "react-datepicker";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";

const ReminderDialog = ({
  open,
  handleClose,
  handleActionButton,
  initialDate,
  closeButtonLabel,
  actionButtonLabel,
}) => {
  const classes = useStyles();

  const [date, setDate] = React.useState(initialDate.toDate());
  const [description, setDescription] = React.useState("");
  const [city, setCity] = React.useState("");

  const handleDatePickerChange = (date) => {
    setDate(date);
  };

  const handleDescriptionChange = (e) => {
    e.preventDefault();
    const text = e.target.value;

    if (text.length <= 30) {
      setDescription(text);
    }
  };

  const handleCityChange = (e) => {
    e.preventDefault();
    const text = e.target.value;

    if (text.length <= 10) {
      setCity(text);
    }
  };

  const handleActionButtonClick = () => {
    if (city && description) {
      const reminder = {
        date: date.toString(),
        description,
        city,
      };
      handleActionButton(reminder);
    }
  };

  return (
    <Dialog
      open={open}
      onClose={handleClose}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">Create a new reminder</DialogTitle>
      <DialogContent>
        <DialogContentText>
          Remember description field will only allow 30 characters max and city
          only 15 characters max.
        </DialogContentText>
        <TextField
          autoFocus
          margin="dense"
          id="description"
          label="description"
          type="email"
          fullWidth
          inputProps={{ maxLength: 30 }}
          onChange={handleDescriptionChange}
          autoComplete="off"
        />
        <TextField
          margin="dense"
          id="city"
          label="City"
          type="email"
          fullWidth
          inputProps={{ maxLength: 15 }}
          onChange={handleCityChange}
          autoComplete="off"
        />
        <Grid container className={classes.gridContainer}>
          <Grid item xs={2}>
            <Typography>Pick a date</Typography>
          </Grid>
          <Grid item xs={4}>
            <DatePicker
              selected={date}
              onChange={handleDatePickerChange}
              showTimeSelect
            />
          </Grid>
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose} color="primary">
          {closeButtonLabel}
        </Button>
        <Button
          onClick={handleActionButtonClick}
          color="primary"
          disabled={!city || !description}
        >
          {actionButtonLabel}
        </Button>
      </DialogActions>
    </Dialog>
  );
};

const useStyles = makeStyles((theme) => ({
  datePickerGrid: {
    marginTop: "-2px",
  },
  gridContainer: {
    padding: theme.spacing(2),
  },
}));

ReminderDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  handleActionButton: PropTypes.func.isRequired,
  initialDate: PropTypes.object.isRequired,
  closeButtonLabel: PropTypes.string.isRequired,
  actionButtonLabel: PropTypes.string.isRequired,
};

export default ReminderDialog;
