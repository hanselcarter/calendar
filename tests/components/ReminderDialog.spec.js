import { shallow } from "enzyme";
import React from "react";
import ReminderDialog from "Components/ReminderDialog";

jest.mock("../../src/services/forecast", () => ({
  fetchForecastByCityName: jest.fn().mockResolvedValue({
    skies: "",
    minTemp: "",
    maxTemp: "",
    temp: "",
    feelsLike: "",
  }),
}));

const saveMock = jest.fn();

describe("<ReminderDialog/>", () => {
  let componentWrapper;
  const initialProps = {
    open: true,
    handleClose: jest.fn(),
    handleActionButton: saveMock,
    initialDate: {
      toDate: jest.fn().mockReturnValue("some date"),
    },
    closeButtonLabel: "close",
    actionButtonLabel: "action",
  };

  beforeEach(() => {
    componentWrapper = shallow(<ReminderDialog {...initialProps} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should match ReminderDialog snapshot", () => {
    expect(componentWrapper).toMatchSnapshot();
  });

  describe("When Text field description changes", () => {
    describe("And Text field  values has more than 30 characters ", () => {
      test("Text field value should not change", () => {
        const longDescription =
          "lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum";
        const descriptionTextField = componentWrapper.find("#description");
        const initialValue = descriptionTextField.prop("value");
        descriptionTextField.simulate("change", {
          preventDefault: jest.fn(),
          target: {
            value: longDescription,
          },
        });

        expect(componentWrapper.find("#description").prop("value")).toBe(
          initialValue
        );
      });
    });

    describe("And Text field  values has less than 30 characters ", () => {
      const shortDescription = "lorem ipsum ";

      beforeEach(() => {
        const descriptionTextField = componentWrapper.find("#description");
        descriptionTextField.simulate("change", {
          preventDefault: jest.fn(),
          target: {
            value: shortDescription,
          },
        });
      });
      test("Text field value should change", () => {
        expect(componentWrapper.find("#description").prop("value")).toBe(
          shortDescription
        );
      });

      describe("And action button gets clicked", () => {
        it("Should start saving the reminder", () => {
          const description = componentWrapper
            .find("#description")
            .prop("value");
          const city = componentWrapper
            .find("#city-simple-select")
            .prop("value");
          const color = componentWrapper
            .find("#color-simple-select")
            .prop("value");

          const actionButton = componentWrapper.find("#action-button");
          actionButton.simulate("click");

          const reminder = {
            date: initialProps.initialDate.toDate(),
            description,
            city,
            color,
          };

          expect(saveMock).toBeCalledWith(reminder);
        });
      });
    });
  });
});
