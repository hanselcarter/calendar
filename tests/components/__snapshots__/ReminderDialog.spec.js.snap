// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReminderDialog/> Should match ReminderDialog snapshot 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="form-dialog-title"
  classes={
    Object {
      "paper": "makeStyles-dialogPaper-7",
    }
  }
  onClose={[MockFunction]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="form-dialog-title"
  >
    Create a new reminder
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogContentText))>
      Remember description field can only have a maximum of 30 characters.
    </WithStyles(ForwardRef(DialogContentText))>
    <WithStyles(ForwardRef(TextField))
      autoComplete="off"
      autoFocus={true}
      fullWidth={true}
      id="description"
      inputProps={
        Object {
          "maxLength": 30,
        }
      }
      label="Description"
      margin="dense"
      onChange={[Function]}
      type="email"
      value=""
    />
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-gridContainer-2"
      container={true}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={4}
      >
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-3"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="city-input-label"
          >
            City
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            id="city-simple-select"
            labelId="city-simple-select-label"
            onChange={[Function]}
            value="London"
          >
            <WithStyles(ForwardRef(MenuItem))
              key="London"
              value="London"
            >
              London
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Tegucigalpa"
              value="Tegucigalpa"
            >
              Tegucigalpa
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Medellin"
              value="Medellin"
            >
              Medellin
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Bogota"
              value="Bogota"
            >
              Bogota
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Shanghai"
              value="Shanghai"
            >
              Shanghai
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Moscow"
              value="Moscow"
            >
              Moscow
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Toronto"
              value="Toronto"
            >
              Toronto
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Madrid"
              value="Madrid"
            >
              Madrid
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Berlin"
              value="Berlin"
            >
              Berlin
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Washington"
              value="Washington"
            >
              Washington
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Tokyo"
              value="Tokyo"
            >
              Tokyo
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="Paris"
              value="Paris"
            >
              Paris
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              key="New York City"
              value="New York City"
            >
              New York City
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={4}
      >
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-3"
        >
          <WithStyles(ForwardRef(InputLabel))
            className="makeStyles-greenMenuItem-5"
            id="color-input-label"
          >
            Color
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            id="color-simple-select"
            labelId="color-simple-select-label"
            onChange={[Function]}
            value="green"
          >
            <WithStyles(ForwardRef(MenuItem))
              className="makeStyles-greenMenuItem-5"
              key="green"
              value="green"
            >
              green
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              className="makeStyles-redMenuItem-4"
              key="red"
              value="red"
            >
              red
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              className="makeStyles-yellowMenuItem-6"
              key="yellow"
              value="yellow"
            >
              yellow
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-gridContainer-2"
      container={true}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={4}
      >
        <WithStyles(ForwardRef(Typography))>
          Pick a date and time:
        </WithStyles(ForwardRef(Typography))>
        <t
          allowSameDay={false}
          customTimeInput={null}
          dateFormat="MM/dd/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          enableTabLoop={true}
          excludeScrollbar={true}
          inlineFocusSelectedMonth={false}
          monthsShown={1}
          nextMonthButtonLabel="Next Month"
          nextYearButtonLabel="Next Year"
          onBlur={[Function]}
          onCalendarClose={[Function]}
          onCalendarOpen={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          previousYearButtonLabel="Previous Year"
          readOnly={false}
          renderDayContents={[Function]}
          selected="some date"
          shouldCloseOnSelect={true}
          showMonthYearPicker={false}
          showPopperArrow={true}
          showPreviousMonths={false}
          showQuarterYearPicker={false}
          showTimeInput={false}
          showTimeSelect={true}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        container={true}
        item={true}
        xs={8}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={12}
        >
          <WithStyles(ForwardRef(Typography))>
            Weather in 
            London
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Skies: 
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Temperature: 
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Max temperature: 
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Min temperature: 
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <WithStyles(ForwardRef(Typography))
            variant="caption"
          >
            Feels like: 
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      id="close-button"
      onClick={[MockFunction]}
    >
      close
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      disabled={true}
      id="action-button"
      onClick={[Function]}
    >
      action
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
